packages/3d-web-voice-chat/src/voice-chat-manager/VoiceChatManager.ts

```diff
@@ -0,0 +1,331 @@
+      }
+      if (this.joinButton!.style.backgroundColor !== "rgba(0, 0, 0, 0.8)") {
+        this.joinButton!.style.backgroundColor = "rgba(0, 0, 0, 0.8)";
+      }
+      if (this.joinButton!.innerHTML !== this.micOnIcon) {
+        this.joinButton!.innerHTML = this.micOnIcon;
+      }
+    }
+  }
+
+  private tick() {
+    if (!this.hasJoinedAudio) return;
+    if (VoxeetSDK.conference.participants.size > 0) {
+      this.setStatusElement();
+      let activeSpeakers = 0;
+      for (const [, participant] of VoxeetSDK.conference.participants) {
+        const parsed = parseInt(participant.info.name!, 10);
+        if (this.debug === true) {
+          console.log("parsed", parsed, "userId", this.userId);
+        }
+        if (!parsed) break;
+        if (participant.status === "Connected" && participant.audioTransmitting === true) {
+          activeSpeakers++;
+        }
+        if (parsed === this.userId) {
+          if (this.latestCharacterObj.characterState) {
+            const { position, rotation } = this.latestCharacterObj.characterState;
+            const eulerRot = getEulerFromQuaternion(rotation);
+            const direction: Direction = { x: eulerRot.pitch, y: eulerRot.yaw, z: eulerRot.roll };
+            if (this.debug === true) {
+              console.log(`ID ${parsed}: ${formatPos(position)}`);
+              console.log(`ID ${parsed}: ${formatDirection(eulerRot)}`);
+            }
+            VoxeetSDK.conference.setSpatialPosition(participant, position);
+            VoxeetSDK.conference.setSpatialDirection(participant, direction);
+          }
+        } else {
+          const remoteUserState = this.remoteUserStates.get(parsed);
+          if (remoteUserState) {
+            const { position, rotation } = remoteUserState;
+            if (this.debug === true) {
+              console.log(`ID ${parsed}: ${formatPos(position)}`);
+            }
+            VoxeetSDK.conference.setSpatialPosition(participant, position);
+          } else {
+            if (this.debug === true) {
+              console.log("No remote user state found for", parsed);
+            }
+            VoxeetSDK.conference.kick(participant);
+          }
+        }
+      }
+      this.activeSpeakers = activeSpeakers;
+    } else {
+      this.setStatusElement();
+    }
+  }
+
+  private bindDOMElements() {
+    this.joinButton = document.getElementById("voice-join-button") as HTMLDivElement;
+    this.joinButton.addEventListener("click", () => {
+      this.joinClick();
+    });
+    this.participantsDiv = document.getElementById("voice-participants-count") as HTMLDivElement;
+  }
+
+  private async init() {
+    try {
+      const accessToken = await fetch(`/voice-token/${this.userId.toString(10)}`)
+        .then((res) => res.json())
+        .then((res) => res.accessToken as string)
+        .catch((err) => {
+          console.error(err);
+          return null;
+        });
+      if (!accessToken) {
+        return;
```

Comment:
The case where voice chat isn't configured (i.e. because API keys weren't specified in the env vars) should probably be handled explicitly and gracefully (e.g. showing audio as unavailable in the UI or hiding the UI)